neuralop/layers/legacy_spectral_convolution.py

```diff
@@ -0,0 +1,722 @@
+    else:
+        return tl.einsum(eq, x, tucker_weight.core, *tucker_weight.factors)
+
+
+def _contract_tt(x, tt_weight, separable=False):
+    order = tl.ndim(x)
+
+    x_syms = list(einsum_symbols[:order])
+    weight_syms = list(x_syms[1:])  # no batch-size
+    if not separable:
+        weight_syms.insert(1, einsum_symbols[order])  # outputs
+        out_syms = list(weight_syms)
+        out_syms[0] = x_syms[0]
+    else:
+        out_syms = list(x_syms)
+    rank_syms = list(einsum_symbols[order + 1 :])
+    tt_syms = []
+    for i, s in enumerate(weight_syms):
+        tt_syms.append([rank_syms[i], s, rank_syms[i + 1]])
+    eq = (
+        "".join(x_syms)
+        + ","
+        + ",".join("".join(f) for f in tt_syms)
+        + "->"
+        + "".join(out_syms)
+    )
+
+    if x.dtype == torch.complex32:
+        return einsum_complexhalf(eq, x, *tt_weight.factors)
+    else:
+        return tl.einsum(eq, x, *tt_weight.factors)
+
+
+def get_contract_fun(weight, implementation="reconstructed", separable=False):
+    """Generic ND implementation of Fourier Spectral Conv contraction
+
+    Parameters
+    ----------
+    weight : tensorly-torch's FactorizedTensor
+    implementation : {'reconstructed', 'factorized'}, default is 'reconstructed'
+        whether to reconstruct the weight and do a forward pass (reconstructed)
+        or contract directly the factors of the factorized weight with the input (factorized)
+    separable : bool
+        whether to use the separable implementation of contraction. This arg is
+        only checked when `implementation=reconstructed`.
+
+    Returns
+    -------
+    function : (x, weight) -> x * weight in Fourier space
+    """
+    if implementation == "reconstructed":
+        if separable:
+            print("SEPARABLE")
+            return _contract_dense_separable
+        else:
+            return _contract_dense
+    elif implementation == "factorized":
+        if torch.is_tensor(weight):
+            return _contract_dense
+        elif isinstance(weight, FactorizedTensor):
+            if weight.name.lower().endswith("dense"):
+                return _contract_dense
+            elif weight.name.lower().endswith("tucker"):
+                return _contract_tucker
+            elif weight.name.lower().endswith("tt"):
+                return _contract_tt
+            elif weight.name.lower().endswith("cp"):
+                return _contract_cp
+            else:
+                raise ValueError(f"Got unexpected factorized weight type {weight.name}")
+        else:
+            raise ValueError(
+                f"Got unexpected weight type of class {weight.__class__.__name__}"
+            )
+    else:
+        raise ValueError(
+            f'Got implementation={implementation}, expected "reconstructed" or "factorized"'
+        )
+
+
+Number = Union[int, float]
+
+
+class SpectralConv(BaseSpectralConv):
```

Comment:
Why do we want to keep this legacy code?

Reply:
In case we want to revert - for specific use cases. That version is harder to read but actually faster in some cases. Some users might want it for that reason